cmd/commands/controller.go

```diff
@@ -0,0 +1,326 @@
+package commands
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+
+	"github.com/spf13/cobra"
+	"go.uber.org/zap"
+	appv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	argoevents "github.com/argoproj/argo-events"
+	"github.com/argoproj/argo-events/common"
+	"github.com/argoproj/argo-events/common/logging"
+	"github.com/argoproj/argo-events/controllers/eventbus"
+	"github.com/argoproj/argo-events/controllers/eventsource"
+	"github.com/argoproj/argo-events/controllers/sensor"
+	eventbusv1alpha1 "github.com/argoproj/argo-events/pkg/apis/eventbus/v1alpha1"
+	eventsourcev1alpha1 "github.com/argoproj/argo-events/pkg/apis/eventsource/v1alpha1"
+	sensorv1alpha1 "github.com/argoproj/argo-events/pkg/apis/sensor/v1alpha1"
+	envpkg "github.com/argoproj/pkg/env"
+)
+
+const (
+	namespaceEnvVar           = "NAMESPACE"
+	natsStreamingEnvVar       = "NATS_STREAMING_IMAGE"
+	natsMetricsExporterEnvVar = "NATS_METRICS_EXPORTER_IMAGE"
+	eventSourceImageEnvVar    = "EVENTSOURCE_IMAGE"
+	sensorImageEnvVar         = "SENSOR_IMAGE"
+
+	componentValEventBus    = "eventbus"
+	componentValEventSource = "eventsource"
+	componentValSensor      = "sensor"
+)
+
+func NewControllerCommand() *cobra.Command {
+	var (
+		component        string
+		namespaced       bool
+		managedNamespace string
+	)
+
+	command := &cobra.Command{
+		Use:   common.ControllerCommand,
+		Short: "Start a controller",
+		Run: func(cmd *cobra.Command, args []string) {
+			switch component {
```

Comment:
This is bothering me, because there are 2 other possible args: `--namespaced` and `--managed-namespace`.

`argo-events controller --compement=eventbus --namespaced --managed-namespace abc`

vs.

`argo-events controller eventbus --namespaced --managed-namespace abc`

Which one is better?

Reply:
These days prefer to use env vars instead of/as well as CLI options.  Kustomize strategy merge patch with them, you cannot easily

`NAMESPACED=true MANAGED_NAMESPACE=abc argo-events controller eventbus`

Thought maybe it should be:

`argo-events event-bus-controller`
`argo-events event-source-controller`
`argo-events sensor-controller`
`NAMESPACED=true argo-events sensor`
`NAMESPACED=true argo-events event-source`

I don't feel super strongly about this, but maybe you can decide which is best for the developer, and what make the code nice.